lib/auth/predicate/builtins.go

```diff
@@ -0,0 +1,361 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package predicate
+
+import (
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+func getIdentifier(obj any, selectors []string) (any, error) {
+	switch selectors[0] {
+	case "false":
+		return false, nil
+	case "true":
+		return true, nil
+	}
+
+	for _, s := range selectors {
+		if obj == nil || reflect.ValueOf(obj).IsNil() {
+			return nil, trace.BadParameter("cannot take field of nil")
+		}
+
+		if m, ok := obj.(map[string]any); ok {
+			obj = m[s]
+			continue
+		}
+
+		val := reflect.ValueOf(obj)
+		ty := reflect.TypeOf(obj)
+		if ty.Kind() == reflect.Interface || ty.Kind() == reflect.Ptr {
+			val = reflect.ValueOf(obj).Elem()
+			ty = val.Type()
+		}
+
+		if ty.Kind() == reflect.Struct {
+			for i := 0; i < ty.NumField(); i++ {
+				tagValue := ty.Field(i).Tag.Get("json")
+				parts := strings.Split(tagValue, ",")
+				if parts[0] == s {
+					obj = val.Field(i).Interface()
+					break
+				}
+			}
+
+			continue
+		}
+
+		return nil, trace.BadParameter("cannot take field of type: %T", obj)
+	}
+
+	return obj, nil
+}
+
+func getProperty(m any, k any) (any, error) {
+	switch mT := m.(type) {
+	case map[string]any:
+		kS, ok := k.(string)
+		if !ok {
+			return nil, trace.BadParameter("unsupported key type: %T", k)
+		}
+
+		return mT[kS], nil
+	default:
+		return nil, trace.BadParameter("cannot take property of type: %T", m)
+	}
+}
+
+func builtinOpAnd(a, b bool) any {
+	return a && b
+}
+
+func builtinOpOr(a, b bool) any {
+	return a || b
+}
+
+func builtinOpNot(a bool) any {
+	return !a
+}
+
+func builtinOpEquals(a, b any) any {
+	return reflect.DeepEqual(a, b)
+}
+
+func builtinOpLT(a, b any) any {
+	if reflect.TypeOf(a) != reflect.TypeOf(b) {
+		return false
+	}
```

Comment:
I think this works, give it a try?
```suggestion
func builtinOpLT(a, b any) (bool, error) {
	if reflect.TypeOf(a) != reflect.TypeOf(b) {
		return false, trace.BadParameter(`args to "<" of types %T and %T do not match`, a, b)
	}
```